### PR TITLE
[PoC] Training Restart - V2

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Different Loop API compared to https://github.com/PyTorchLightning/pytorch-lightning/pull/8131

TODOS
[] Resolve `batch_idx` not properly reset in loops
[] Add ProgressBar proper indices
[] Resolve _map_dl_idx_sampler_states should be updated after the batch is processed
[] Save extras for on epoch_end
[] Add ResultCollection to `state dict` and `reload` automatically on `Base Loop`
[] Enable non-fault tolerant reload.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [ ] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/base.py [hunk 1]
```diff
@@ -85,12 +172,18 @@ def run(self, *args: Any, **kwargs: Any) -> Optional[Any]:
...
             return self.on_skip()
 
         if self.restarting:
-            self.restore()
-            self.restarting = False
+            if not is_overridden("restore", self, Loop):
+                warning_cache.warn(f"{self.__class__.__name__} Loop doesn't override the restore function.")
+            is_restore_finished = self.restore()
+            if is_restore_finished is None or is_restore_finished:
+                self.restarting = False
```
Comment (awaelchli) [inline pytorch_lightning/loops/base.py:186]: instead of 
```
if restarting
   self.restore()
else:
    self.reset()
```

would it make sense to do:

```
self.reset()
if restarting:
    self.restore()

```
?

Then we don't need the `self._initialize()` in the subclassed loops and we can assume after `reset()` the state is properly initialized and we can restore if appropritae.

File: pytorch_lightning/loops/base.py [hunk 2]
```diff
@@ -108,8 +201,13 @@ def run(self, *args: Any, **kwargs: Any) -> Optional[Any]:
...
         return output
 
-    def restore(self) -> None:
-        """Restore the internal state of the loop the beginning of run if restarting is ``True``."""
+    def restore(self) -> Optional[bool]:
+        """
+        Restore the internal state of the loop the beginning of run if restarting is ``True``.
+
+        Returns:
+            is_restore_finished: Whether the restoration actually finished.
```
Comment (awaelchli) [inline pytorch_lightning/loops/base.py:209]: in which case will a restoration not finish. 
is this this about a success/fail return signal?

File: pytorch_lightning/loops/base.py [hunk 3]
```diff
@@ -142,9 +240,62 @@ def on_run_end(self) -> Any:
...
-
     def state_dict(self) -> Dict:
-        """Return the loop current states."""
+        """Current Loop state"""
         return {}
+
+    def load_state_dict(self, state_dict: Dict) -> None:
+        """Reload Loop state"""
+
+    def get_state_dict(self, destination: Optional[OrderedDict] = None, prefix: Optional[str] = '') -> OrderedDict:
```
Comment (awaelchli) [inline pytorch_lightning/loops/base.py:250]: as discussed, could we consider merging several of these state_dict related APIs into one. 
IMO we should aim at keeping the base loop class as simple as possible.

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 4]
```diff
@@ -112,14 +107,28 @@ def run(self, batch: Any, batch_idx: int, dataloader_idx: int) -> AttributeDict:
...
+        self._initialize()
+
+        # restarting isn't finished yet.
+        return False
+
+    def reset(self) -> None:
+        """Resets the loop state"""
+        self._initialize()
+
+        self.optim_progress.optimizer.reset_on_epoch()
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:130]: these _initialize() calls could be simplified as proposed in the earlier comment

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 5]
```diff
@@ -131,6 +140,14 @@ def on_run_start(self, batch: Any, batch_idx: int, dataloader_idx: int):
         void(batch_idx, dataloader_idx)
         self._remaining_splits = list(enumerate(self._tbptt_split_batch(batch)))
 
+    def on_advance_start(self, *args: Any, **kwargs: Any) -> None:
+        self.progress.increment_started()
+        return super().on_advance_start(*args, **kwargs)
+
+    def on_advance_end(self) -> None:
+        self.progress.increment_completed()
+        return super().on_advance_end()
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:149]: we should probably increment after the super call?

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 6]
```diff
@@ -148,7 +165,18 @@ def advance(self, batch, batch_idx, dataloader_idx):
...
         if self.trainer.lightning_module.automatic_optimization:
-            for opt_idx, optimizer in self.get_active_optimizers(batch_idx):
+            active_optimizers = self.get_active_optimizers(batch_idx)
+            for opt_idx, optimizer in active_optimizers:
+
+                # handle optimization restart
+                if self.restarting:
+                    if len(active_optimizers) > 1 and opt_idx < self.progress.current.completed:
+                        continue
+                    self.restarting = False
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:175]: okay I see.
this is something a optimizer loop could handle for restarting.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 7]
```diff
@@ -54,37 +54,35 @@ def __init__(self, min_steps: int, max_steps: int):
...
+        return self.progress.batch.current.completed
+
     @property
     def done(self) -> bool:
         """Returns whether the training should be stopped.
         The criteria are that the number of steps reached the max steps,
         the last batch is reached or the trainer signals to stop (e.g. by early stopping).
         """
-        max_steps_reached = self.max_steps is not None and self.global_step >= self.max_steps
+        max_steps_reached = self.max_steps is not None and (self.total_optimizer_step) >= self.max_steps
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:82]: max steps in the trainer does not represent optimizer steps (today). It represents how many times training_step() was called (roughly). If we want to change that I think it should be done in a separate PR with a proper evaluation of what the effects are.

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 8]
```diff
@@ -165,6 +165,17 @@ def setup_environment(self) -> None:
 
         self.setup_distributed()
 
+        # share ddp pids to all processes
+        self.share_pids()
+
+    def share_pids(self):
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp.py:171]: leftover from a merge :)

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 9]
```diff
@@ -141,6 +144,13 @@ def restore_model(self) -> None:
         # restore model state_dict
         self.trainer.training_type_plugin.load_model_state_dict(self._loaded_checkpoint)
 
+        gradients = self._loaded_checkpoint.get("gradients", None)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:147]: what would we call this feature? gradient checkpointing 🤣

File: pytorch_lightning/trainer/connectors/loop_connector.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+class LoopConnector(object):
+
+    def __init__(self, trainer: "pl.Trainer"):
+        self.trainer = trainer
+
+    def on_trainer_init(self, *args, **kwargs):
+        self.trainer.fit_loop = self.on_fit_loop_init(*args, **kwargs)
+        self.trainer.validate_loop = self.on_validate_loop_init(*args, **kwargs)
+        self.trainer.test_loop = self.on_test_loop_init(*args, **kwargs)
+        self.trainer.predict_loop = self.on_predict_loop_init(*args, **kwargs)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/loop_connector.py:30]: a loop connector could be useful in general, but we said we wanted to avoid Trainer state management outside the trainer. If Trainer owns loops, imo it would be safer to let the Trainer manage its state.

cc @ananthsub @justusschock

**Review by awaelchli — COMMENTED** [review]
review part 1/n

this PR is integrating progress tracking
do you see a way to break it out or is it rather not so easy?

File: pytorch_lightning/trainer/data_loading.py [hunk 11]
```diff
@@ -248,23 +310,40 @@ def reset_train_dataloader(self, model: 'pl.LightningModule') -> None:
...
+
+        # detect iterable dataset
+        contains_iterable_dataset = False
+
+        def detect_iterable_dataset(dataloader: DataLoader):
+            nonlocal contains_iterable_dataset
+            if isinstance(dataloader.dataset, IterableDataset):
+                contains_iterable_dataset = True
+
+        apply_to_collection(loaders, DataLoader, detect_iterable_dataset)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/data_loading.py:325]: could potentially be factored out into a helper function.

File: pytorch_lightning/trainer/progress.py [hunk 12]
```diff
@@ -25,14 +25,23 @@ def load_state_dict(self, state_dict: dict) -> None:
...
 
     @classmethod
-    def from_state_dict(cls, state_dict: dict) -> "_DataclassStateDictMixin":
+    def from_state_dict(cls, state_dict: dict) -> "BaseProgress":
         obj = cls()
         obj.load_state_dict(state_dict)
         return obj
 
 
+class ProgressDict(Dict):
```
Comment (awaelchli) [inline pytorch_lightning/trainer/progress.py:34]: isn't this what a regular dict can do?

```py
old_dict = dict(...)
new_dict = dict(old_dict)
new_dict.update(...)
```

File: pytorch_lightning/trainer/supporters.py [hunk 13]
```diff
@@ -371,6 +373,62 @@ def __init__(self, loaders: Any, mode: str = 'min_size'):
...
+                fast_forward_sampler_load_state_dict,
+            )
+
+            count = 0
+            apply_to_collection(
+                self._iterator.loader_iters, Iterator,
+                partial(cycle_to_next_worker, state_dict=self.loaders_iter_state_dict, count=count)
+            )
+
+            count = 0
```
Comment (awaelchli) [inline pytorch_lightning/trainer/supporters.py:426]: these local vars can be removed, right?

File: pytorch_lightning/trainer/supporters.py [hunk 14]
```diff
@@ -514,14 +574,18 @@ def _nested_calc_num_data(data: Union[Mapping, Sequence], compute_func: Callable
...
+                      trainer: Optional['pl.Trainer'] = None) -> Generator[Tuple[Any, bool], None, None]:
     """
     Returns an iterator that pre-fetches and caches the next item.
     The values are passed through from the given iterable with an added boolean indicating if this is the last item.
     See `https://stackoverflow.com/a/1630350 <https://stackoverflow.com/a/1630350>`_
     """
     it = iter(iterable)
 
+    if trainer:
+        trainer.current_iterator = it
```
Comment (awaelchli) [inline pytorch_lightning/trainer/supporters.py:587]: can you explain what this workaround is about?
I can see that you are using it in the checkpoint connector but it is not clear to me what problem it is trying to solve.

File: pytorch_lightning/trainer/trainer.py [hunk 15]
```diff
@@ -995,12 +995,41 @@ def _run_train(self) -> None:
             if distributed_available() and self.world_size > 1:
                 # try syncing remaing processes, kill otherwise
                 self.training_type_plugin.reconciliate_processes(traceback.format_exc())
+            # save a checkpoint for fault tolerant training
+            self.fit_loop._check_checkpoint_callback(True)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:999]: @carmocca that's what I meant the other day when we talked about why we have the `check_checkpoint_callback` in the first place. 

this usage here was the old way Lightning would save checkpoints on interrupt.

File: pytorch_lightning/utilities/auto_restart.py [hunk 16]
```diff
@@ -219,17 +222,140 @@ def __next__(self) -> Dict[str, Any]:
...
+                    for sampler_name, sampler_state_dict in v.items():
+                        iterable_dataset_state_dict[sampler_name] = {
+                            worker_id: {
+                                "current_iteration": sampler_state_dict[worker_id]["current_iteration"][-1].item()
+                            }
+                        }
+                    state_dict.append(iterable_dataset_state_dict)
+                    return data["data"]
+                out.append((k, CaptureIterableDataset._sanetize_batch_from_sampler_state(v, state_dict)))
+            return elem_type(OrderedDict(out))
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:245]: is there a specific reason for using `OrderedDict` vs. just dict?

File: pytorch_lightning/trainer/trainer.py (hunk 15 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:999]: We shouldnt manually call the `ModelCheckpoint` callback like this as it will be moved to a fully hook-based callback next week.

If training is stopped and we need to save, we should do it by calling the `on_train_end` hook

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:999]: Adding to this:

If we need to manually save a checkpoint, we can always call `trainer.save_checkpoint()` and this doesn't go through the ModelCheckpoint :)

File: pytorch_lightning/loops/base.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:209]: This is just a mechanism to enable a loop to delay the restarting variable to be set to False.

In `TrainingBatchLoop`, this property is required to know if we should `fast-forward` the dataloaders, therefore by return False, I indicate to the loop I will manually switch off the restarting boolean later on in the loop.

Example:
https://github.com/PyTorchLightning/pytorch-lightning/pull/8337/files#diff-eba04421e2e60c9d7b55f56ca57c9a319a4e3de38e3eb96c931324ecacbb73e0R175

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:149]: Done !

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:175]: Yes, exactly !

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:147]: If the users had gradient accumulation activated.

File: pytorch_lightning/utilities/auto_restart.py (hunk 16 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:244]: Following apply_to_collection implementation. I believe there were a reason why OrderedDict was selected over it.

File: pytorch_lightning/trainer/supporters.py (hunk 14 above)
Comment (tchaton) [reply pytorch_lightning/trainer/supporters.py:600]: This can be removed ! This was used to not lose the Iterator reference as it contains the DataLoader Iterators.

File: pytorch_lightning/trainer/trainer.py (hunk 15 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:999]: I would prefer for the ModelCheckpoint to save the checkpoint as the last_model_path should be the one saved on failure.
